core/drivers/nvmem/nvmem_die_id.c

```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, Microchip
+ */
+
+#include <drivers/nvmem.h>
+#include <io.h>
+#include <kernel/dt.h>
+#include <kernel/huk_subkey.h>
+#include <kernel/tee_common_otp.h>
+#include <malloc.h>
+#include <matrix.h>
+#include <sama5d2.h>
+#include <tee_api_defines.h>
+#include <tee_api_types.h>
+#include <types_ext.h>
+#include <trace.h>
+
+static uint8_t *die_id;
+static size_t die_id_len;
+
+int tee_otp_get_die_id(uint8_t *buffer, size_t len)
+{
+	if (!die_id) {
+		if (huk_subkey_derive(HUK_SUBKEY_DIE_ID, NULL, 0, buffer, len))
+			return -1;
+		return 0;
+	}
+
+	memcpy(buffer, die_id, MIN(die_id_len, len));
+
+	return 0;
+}
+
+static TEE_Result nvmem_die_id_read(const void *fdt, int node)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct nvmem_cell *cell = NULL;
+	uint8_t *data = NULL;
+
+	res = nvmem_get_cell_by_name(fdt, node, "die_id", &cell);
+	if (res)
+		return res;
+
+	res = nvmem_cell_malloc_and_read(cell, &data);
+	if (!res)
+		die_id = data;
+
+	nvmem_put_cell(cell);
+
+	return res;
+}
+
+static TEE_Result nvmem_die_id_probe(const void *fdt, int node,
+				     const void *compat_data __unused)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	return res = nvmem_die_id_read(fdt, node);
```

Comment:
can remove completely the res local variable and
 `return nvmem_die_id_read(fdt, node);`

I'd even suggest to move what is done in nvmem_die_id_read() in the probe function since this is the only thing done in the probe for now but that's not a strong opinion